security/pkg/nodeagent/caclient/providers/citadel/client.go

```diff
@@ -127,13 +130,24 @@ func (c *CitadelClient) getTLSDialOption() (grpc.DialOption, error) {
 				certificate, err = tls.LoadX509KeyPair(
 					filepath.Join(c.opts.ProvCert, "cert-chain.pem"),
 					filepath.Join(c.opts.ProvCert, "key.pem"))
+
 				if err != nil {
 					// we will return an empty cert so that when user sets the Prov cert path
 					// but not have such cert in the file path we use the token to provide verification
 					// instead of just broken the workflow
 					citadelClientLog.Warnf("cannot load key pair, using token instead: %v", err)
 					return &certificate, nil
 				}
+				var isExpired bool
+				isExpired, err = c.isCertExpired(filepath.Join(c.opts.ProvCert, "cert-chain.pem"))
+				if err != nil {
+					citadelClientLog.Warnf("cannot parse the cert chain, using token instead: %v", err)
+					return &tls.Certificate{}, nil
+				}
+				if isExpired {
+					citadelClientLog.Warnf("cert expired, using token instead")
```

Comment:
I think that Istiod should allow mTLS cert *or* token. Its not the clients responsibility to ensure it sends only valid credentials, Istiod should be modified to properly use OR for auth, instead of short-circuiting when the TLS cert is invalid

Reply:
So the logic here just add that. if the certificate is expire we will check the token instead

Reply:
I mean on the server side in Istiod. This is just the client side

Reply:
server side logic is much difficult to change. As it will only check one of the token or cert. And in our original logic, we will only attach the token when the cert is empty in the CSR

Reply:
I think its more correct behavior, it should not be much more complex, we just need to modify the TLs config on the server (https://github.com/istio/istio/blob/cd4e73cd8fdca373d89ae1b8175a1b0c2d4d0658/pilot/pkg/bootstrap/server.go#L749) instead of on the client.

Authentication already checks for one of the auth methods working (not all): https://github.com/istio-testing/istio/blob/6fa684aaef46264a68d4ef377b2bbb433bc87971/pilot/pkg/xds/authentication.go#L56. The problem is we short-circuit at the TLS level

Reply:
I think Istiod is already validating cert and reject it if cert is expired. This PR is to fix client side behavior. With valid token and expired cert, token should be used instead. John, are you suggesting that client should always send all credentials and let Istiod decide which one to use?

Reply:
Yes

Reply:
This sounds good to me. Thanks.

Reply:
I have just checked the code howard mention
Current the failure for this is our tls mode choose the 
`VerifyClientCertIfGiven` Type
https://github.com/williamaronli/istio/blob/cd4e73cd8fdca373d89ae1b8175a1b0c2d4d0658/pilot/pkg/bootstrap/server.go#L749

And tls failed when they receive an expired cert according to:

```
// VerifyClientCertIfGiven indicates that a client certificate should be requested
	// during the handshake, but does not require that the client sends a
	// certificate. If the client does send a certificate it is required to be
	// valid.
	VerifyClientCertIfGiven
```
So when client side send a expired cert to it the tls will fail.


So there may be two solutions:
1. I change it to other options like :
```
// RequireAnyClientCert indicates that a client certificate should be requested
	// during the handshake, and that at least one certificate is required to be
	// sent by the client, but that certificate is not required to be valid.
	RequireAnyClientCert
```
So if changes this tls will not verify the cert, our authentication function will verify it, as right now the csr will contain both token and cert
https://github.com/istio-testing/istio/blob/6fa684aaef46264a68d4ef377b2bbb433bc87971/pilot/pkg/xds/authentication.go#L56

however this may change the security level on server side

2. Another solution is just like this PR:
We check the cert's validation in client side and if it is not valid we don't include that cert in the CSR

@howardjohn is my understanding correct?